Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx

```diff
@@ -0,0 +1,261 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+///
+/// \file    runFileReader.cxx
+/// \author  Andrea Ferrero
+///
+/// \brief This is an executable that reads a data file from disk and sends the data to QC via DPL.
+///
+/// This is an executable that reads a data file from disk and sends the data to QC via the Data Processing Layer.
+/// It can be used as a data source for QC development. For example, one can do:
+/// \code{.sh}
+/// o2-qc-run-file-reader --infile=some_data_file | o2-qc --config json://${QUALITYCONTROL_ROOT}/etc/your_config.json
+/// \endcode
+///
+
+#include <random>
+#include <iostream>
+#include <fstream>
+#include <stdexcept>
+#include "Framework/CallbackService.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/ControlService.h"
+#include "Framework/DataProcessorSpec.h"
+#include "Framework/Lifetime.h"
+#include "Framework/Output.h"
+#include "Framework/Task.h"
+#include "Framework/WorkflowSpec.h"
+
+#include "DPLUtils/DPLRawParser.h"
+#include "MCHBase/Digit.h"
+#include "Headers/RAWDataHeader.h"
+#include "MCHRawCommon/DataFormats.h"
+#include "MCHRawDecoder/PageDecoder.h"
+#include "MCHRawElecMap/Mapper.h"
+#include "MCHMappingFactory/CreateSegmentation.h"
+#include "MCHRawCommon/RDHManip.h"
+
+#include "MCHWorkflow/DataDecoderSpec.h"
+//#include "RawBufferDecoder.h"
+
+// Dans ce code, on récupère un infut aui est un message avec le buffer, on fait tourner le code de base decodeBuffer qui est dans Handlers, et on renvoir un message de sortie (inspiré de FileReader de Andrea)
+
+
+
+namespace o2::header
+{
+extern std::ostream& operator<<(std::ostream&, const o2::header::RAWDataHeaderV4&);
+}
+
+
+
+namespace o2
+{
+namespace mch
+{
+namespace raw
+{
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::mch::mapping;
+using RDHv4 = o2::header::RAWDataHeaderV4;
+
+
+std::array<int, 64> refManu2ds_st345 = {
+    63, 62, 61, 60, 59, 57, 56, 53, 51, 50, 47, 45, 44, 41, 38, 35,
+    36, 33, 34, 37, 32, 39, 40, 42, 43, 46, 48, 49, 52, 54, 55, 58,
+    7, 8, 5, 2, 6, 1, 3, 0, 4, 9, 10, 15, 17, 18, 22, 25,
+    31, 30, 29, 28, 27, 26, 24, 23, 20, 21, 16, 19, 12, 14, 11, 13};
+
+int manu2ds(int i){
+  return refManu2ds_st345[i];
+}
+
+
+class DataDecoderTask
+{
+  void decodeBuffer(gsl::span<const std::byte> page, std::vector<o2::mch::Digit> &digits)
+  {
+    size_t ndigits{0};
+
+    auto channelHandler = [&](DsElecId dsElecId, uint8_t channel, o2::mch::raw::SampaCluster sc) {
+      auto s = asString(dsElecId);
```

Comment:
for performance reasons, this debug string should only be computed where needed (in the mPrint if below)

Reply:
Done.